cli/azd/pkg/commands/pipeline/pipeline.go

```diff
@@ -0,0 +1,94 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package pipeline
+
+import (
+	"context"
+	"encoding/json"
+
+	"github.com/azure/azure-dev/cli/azd/internal"
+	"github.com/azure/azure-dev/cli/azd/pkg/environment"
+	"github.com/azure/azure-dev/cli/azd/pkg/environment/azdcontext"
+	"github.com/azure/azure-dev/cli/azd/pkg/input"
+	"github.com/azure/azure-dev/cli/azd/pkg/tools"
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+)
+
+type subareaProvider interface {
+	requiredTools() []tools.ExternalTool
+	preConfigureCheck(ctx context.Context, console input.Console) error
+	name() string
+}
+
+type gitRepositoryDetails struct {
+	owner          string
+	repoName       string
+	gitProjectPath string
+}
+
+type scmProvider interface {
+	subareaProvider
+	gitRepoDetails(ctx context.Context, remoteUrl string) (*gitRepositoryDetails, error)
+	// configureGitRemote returns the git repository url after setting it
+	configureGitRemote(ctx context.Context, repoPath string, remoteName string, console input.Console) (string, error)
+	preventGitPush(
+		ctx context.Context,
+		gitRepo *gitRepositoryDetails,
+		remoteName string,
+		branchName string,
+		console input.Console) (bool, error)
+}
+
+type ciProvider interface {
+	subareaProvider
+	configurePipeline(ctx context.Context) error
+	configureConnection(
+		ctx context.Context,
+		azdEnvironment environment.Environment,
+		gitRepo *gitRepositoryDetails,
+		credential json.RawMessage) error
+}
```

Comment:
Feels wrong to have these provider interfaces within the same package as the commands themselves...

Reply:
Today we only have a use case to use these providers in the context of a pipeline config, but it would benefit from making these providers public so that they could more easily reused in other commands in the future.

Reply:
I would wait until we know where else we would want to use the interfaces. They are currently designed for CI/CD pipeline configuration specific